### ISSUE TITLE
depends_on cannot be used in a module

### ISSUE THREAD

**Issue by chrisrlong** [hashicorp/terraform#10462]
Hi there,

### Terraform Version
0.8.0 rc1+

### Affected Resource(s)
module

### Terraform Configuration Files
```hcl
module "legacy_site" {
  source = "../../../../../modules/site"
  name = "foo-site"
  health_check_target = "TCP:443"
  azs = "${var.azs}"
  instance_count = "${var.instance_count}"
  vpc = "apps"
  region = "${var.region}"
  environment = "${var.environment}"
  run_list = "hs_site_foo"

  #rds_complete = "${module.rds.db_instance_id}"
  #elasticache_cache_complete = "${module.elasticache_cache.elasticache_id}"
  #elasticache_sessions_complete = "${module.elasticache_sessions.elasticache_id}"

  depends_on = [
  "module.rds",
  "module.elasticache_sessions"
  ]

}
```

### Debug Output
Error loading Terraform: module root: module legacy_site: depends_on is not a valid parameter
module root: module legacy_site: depends_on is not a valid parameter

### Expected Behavior
I am trying to use the new depends_on instead of the above outputs, so I create and provision my app once I know database and caches are built.

### Actual Behavior
Nothing as terraform errors out as above.

### Steps to Reproduce

1. `terraform apply`

### References
depends_on can reference modules. This allows a resource or output to depend on everything within a module. (#10076)

**Comment by stack72** [comment]
Hi @chrisrlong 

This is very strange - we introduced depends_on for modules in 0.8.0-beta2

can you run terraform version and post the output here for me?

Paul

**Comment by chrisrlong** [comment]
Hi @stack72,

I have complied master from version b15b8bd (Terraform v0.8.0-dev (b15b8bd99aae33c5b68cbecf5aef375c0798147c+CHANGES)), which should include beta2 features. (I did this to get 10337  and 10338 fixes)

It seems it would work if you set a resource to depend on a module, but you cannot set a module to depend_on another module......

Thanks for the great tool btw ;)

Chris

**Comment by stack72** [comment]
Hi @chrisrlong 

thanks for getting back to me - you are indeed correct - modules cannot (currently) depend_on other modules but can depend on a resource only

Will change the tag on this from bug to enhancement :)

Keep an eye out for it soon ™️ 

Thanks

Paul

**Comment by mitchellh** [comment]
Retagging as enhancement, since this isn't broken functionality, its functionality that doesn't exist yet. :)

**Comment by sheerun** [comment]
My use case: the slave virtual machine depends on master virtual machine to exist

**Comment by shantanugadgil** [comment]
I am seeing the same error:
`module root: module example: depends_on is not a valid parameter`

My Terraform version is:
Terraform v0.8.2

My use case is that I want a module to depend on a resource.

**Comment by arehmandev** [comment]
Error loading Terraform: module root: module etcdbastion: depends_on is not a valid parameter.
^ Seeing it here too

**Comment by cemo** [comment]
@mitchellh I believe this one can solve unnested modules problem too. There were some issues regarding this #10883 before. This solution would be intuitive and address these issues.

**Comment by fieryvova** [comment]
Waiting for this feature to be added :+1:

**Comment by pseudobeer** [comment]
+1

**Comment by avatar4d** [comment]
+1

**Comment by panmanphil** [comment]
+1 This would really help modularizing our code

**Comment by shoelessrob** [comment]
+1

**Comment by mschenck** [comment]
+1

**Comment by cpanayides** [comment]
+1

**Comment by derBroBro** [comment]
+1

**Comment by waffleshop** [comment]
+1

**Comment by laiph** [comment]
+1

**Comment by mqasim1983** [comment]
+1

**Comment by seyeda** [comment]
+1

**Comment by AirbornePorcine** [comment]
Anyone have a good workaround for this in the meantime?

**Comment by jacobwgillespie** [comment]
My pseudo workaround is to add a list variable inside the module:

```hcl
variable "depends_on" { default = [], type = "list" }
```

Then when using the module, pass it a computed value from the resource I want it to depend on:

```hcl
module "something" {
  depends_on = ["${aws_instance.instance.private_ip}"]
}
```

**Comment by AirbornePorcine** [comment]
Thanks, that doesn't entirely fit what I think this issue is talking about though - making a module able to depend on another one. I don't have any resources that the module should depend on - it should depend on another module being created.

I've tried doing stuff like this:
module 1:
`
output "wait_for_cluster" { value = "ref to a resource that gets created last by this module" }
`

module 2:
`
variable "wait_for_cluster" {}
`

Main terraform template:
`
module "1" {

}

module "2" {
 wait_for_cluster = "${module.1.wait_for_cluster}"
}
`

But this doesn't do anything - my module two is still created at basically the same time as module 1.

**Comment by b-dean** [comment]
@AirbornePorcine In your second module you need to actually use the wait_for_cluster variable somewhere that establishes a dependency. Such as in a template or a trigger on a `null_resource`, etc.

See https://github.com/hashicorp/terraform/issues/1178#issuecomment-207369534 from @kristjanelias

The only change I'd make to his workaround is to maybe make the dummy dependency resource use a trigger so it changes when the instance changes.

```hcl
resource "null_resource" "dummy_dependency" {
  triggers {
    dependency_id = "${aws_instance.instance.id}"
  }
}
```

**Comment by AirbornePorcine** [comment]
Ah, that makes sense! Should have known that Terraform would optimize away my variable if I didn't use it. :) I tried out your suggestion with the trigger and that didn't do it, but setting the depends_id on a tag on some resource in module 2 seems to have done it. Hacky but oh well I guess.

**Comment by eedwardsdisco** [comment]
@stack72 

On 0.9.0

Still getting "depends_on is not a valid parameter" when used within a module definition. I see a lot of +1s asking to be able to depend on another module itself, but the feature that's supposed to work (depend on a module's resource), is broken.

This is biting me HARD because of the big warning at the top of the [ecs service page](https://www.terraform.io/docs/providers/aws/r/ecs_service.html) which describes needing to use depends_on to prevent a race condition. Due to this bug, **you cannot prevent it**.

**Comment by NickMetz** [comment]
+1

**Comment by sbatchu0108** [comment]
+1

**Comment by awebneck** [comment]
+1

**Comment by artrunde** [comment]
+1

**Comment by gustavosoares** [comment]
+1

**Comment by Wenzil** [comment]
+1

**Comment by aholbreich** [comment]
+1

**Comment by BalajiVM** [comment]
-1 
in our use case, instance in the private subnet has to be create after the NAT gateway is created. But, here instances are launching prior to the NAT gateway because of that user data is not executing.

**Comment by jverducc** [comment]
+1

**Comment by jacobwgillespie** [comment]
**Please do not add a +1 comment to this issue.**  Instead, please add a 👍 reaction to the issue parent above.

Many of us are subscribed to the comments here as we are waiting for an update if/when this becomes solved or a workaround is found.  Every "+1" comment triggers another email, essentially spamming the repo owners *and* everybody subscribed.  We are all anxious to find a solution here, but please be mindful of your fellow devs and do not add any additional "+1" comments as they do not work towards finding a solution (you can continue to voice support with a 👍 above).

**Comment by mudrii** [comment]
work around to use -target if you have dependency in module that uses  count = "${length(var.some_var_generated)}"

A way to fix it is to use shell script and trigger missing decencies ex belwo 3 modules 1 vpn creation 2 subnets and 3rd is routing  run vpc and subnets before route module

#!/bin/sh -v
terraform plan -out=tfplan -target=module.vpc -target=module.subnets && \
terraform apply tfplan && \
terraform plan -out=tfplan && \
terraform apply tfplan

**Comment by mattford63** [comment]
An alternate work-around, for my own use-case, is to set a bool explicitly so that the length need not be calculated at the initial planning phase.

**Comment by SqiSch** [comment]
+1

**Comment by ketzacoatl** [comment]
I think #12570 and similar issues would be addressed by this enhancement. Right?

**Comment by dtserekhman-starz** [comment]
Hi, any ETA of when this issue fix/enhancement may be available?

Terraform modules is a powerful feature for abstraction and code-reuse, but without inter-modules dependencies support we can only go so far with this feature. It's blocking us right now for the cases described above.

Thank you.

**Comment by Jonesie** [comment]
This is still failing in 0.9.10.  :/

**Comment by apparentlymart** [comment]
Hi all! Thanks for the great discussion here.

I wanted to share a little background on what's going on here to help explain why Terraform behaves the way it does today and why this is not a simple thing to implement:

Terraform modules today are an organizational/namespacing construct for configuration but they are _not_ actually part of the graph. Rather than the module _itself_ appearing in the graph, you instead find the individual variables and outputs as separate graph nodes, which for _many_ cases is the better model since it means that parts of the calling module can start to get created as soon as their required outputs are finished, rather than waiting until the entire module completes.

The fact that modules themselves don't exist in the graph is why `depends_on` isn't supported for modules. Instead, any interpolation used in either a variable passed into a module or in an output returned from a module creates an implied dependency with _that specific variable or module_.

So with all of this said, I understand that the current behavior is proving inconvenient for some of you in certain cases. If you'd be willing, I'd like to learn more about what these situations are, since there may be other solutions that can address specific use-cases in a way that would be less complex to implement and thus more likely to get done sooner. If any of you are able to share some real-world examples of situations where you've needed a whole-module dependency (that is, where dependency on specific outputs is insufficient) -- and, ideally, some configuration examples illustrating those -- this would be really helpful in figuring out what change is best to make here to address these use-cases.

Thanks again for the discussion here and sorry this issue has been quiet for so long.

**Comment by cemo** [comment]
@apparentlymart I have seen many times, in order to depend on multiple resources from a module, developers create a `null resource` and export it as `depends-id`.  Than other modules are taking input variable as this `depends-id` resource. This process is pretty much ugly and not readable at all. 

I can see current architecture support modules as first class citizen in the graph. I am not an expert for internals but I have an idea:

1. Create an implicit hidden resource for each module. 
2. This implicit resource must be depends on all resources on the modules.
3. Resources in the module which depends on other module can simply implicitly `depends_on` resources which declared at (1) .

Is this doable? What do you think @apparentlymart ?

**Comment by chrisrlong** [comment]
Hi @apparentlymart,

For my case I have a module for specific groupings of components. I originally raised this ticket so that I could guarantee the data (RDS) module and a cache (elasticache) module was finished before launching the site (EC2) module, as the site module fails to complete, [null resource does a deployment to the EC2 instances], due to not have a database/cache to connect to.

I resolved this by making an output for each module and then pass the output to the site module "rds_complete = module.rds.instance_id" using the mechanism you describe.

**Comment by jacobwgillespie** [comment]
@apparentlymart - for my use-case I wanted to make an entire module dependent on a single resource, not depend on an entire module's resources.

I had a module that launched a cluster of various different EC2 clusters/ASGs that, when launched, would run a particular application.  I wanted to make the entire module depend on the existence of an RDS database that would be created outside the module.  I suppose I could pass in a computed value from the RDS instance as a variable and then reference that variable in some dummy way for the EC2 ASGs, but it would be convenient to be able to specify `depends_on` on the module itself and have all of the resources contained within depend on the external resource.

**Comment by FlorinAndrei** [comment]
@apparentlymart Sure, but that's an explanation from a low level perspective. That's just the implementation. You're saying "this car cannot make right-hand turns because we've decided to use chrome-plated thingamajiggers in the steering linkage".

The way it looks for a user, you're creating an abstraction (the module) that behaves in many ways like other entities that can use `depends_on`, but then this abstraction turns around and makes itself an exception to that rule. It's jarring. It really seems to me like this should have been taken into account from the beginning.

**Comment by apparentlymart** [comment]
Thanks for sharing your use-cases, everyone. That helps a lot.

I wanted to note that the `depends_on` mechanism is intended not as the primary means of dealing with dependencies but rather as a workaround for situations where dependencies cannot be inferred automatically. For example, the resources for Amazon API Gateway have some issues here because there are some restrictions on what order resources can be created in that aren't represented in the resource attributes themselves. It's always preferable to use interpolation to create dependencies where it is possible to do so, since it's a more direct description of the desired end-state, rather than of the details of how to get there.

So if possible I'd like to collect some more use-cases where interpolation _can't_ solve the problem via implied dependencies. @chrisrlong's use-case seems like it could be an example of this: the dependency between the application's compute resources and its data store are presumably not explicit in the attributes; I assume in this case that some other configuration store like Consul is acting as the glue to help the code running on the EC2 instance find the RDS and ElastiCache addresses?

@FlorinAndrei I understand and respect that perspective. I wasn't explaining how it works as an excuse for not supporting these use-cases but more as context for my later request that we talk about use cases _rather than_ implementation strategy in case there's a different way to get the outcomes we need here without losing the benefits of the current implementation. I _do_ want to solve this, but as engineers many of us have a tendency to focus on a particular solution to a problem rather than digging into the problem itself, and I want to be open to other solutions that may allow this problem to get solved sooner or with fewer drawbacks.

**Comment by dtserekhman-starz** [comment]
Hi @apparentlymart ,

Thank you for the explanation. I sensed it would require some major redesign/refactoring. Here's my use case:

What I'm trying to do is wrap some main resource creation, such as aws_security_group, etc.. around modules of their own so I can enforce special naming conventions and other default behaviors pertaining to our environment so that other developers can then use these modules as building blocks to create/provision their infrastructure components.

Hence, I have a module that wraps around aws_elb resource instantiation. But then this module requires security_groups dependency. I then have a module that creates a security_group with our naming conventions, etc... And after that, I'm passing in module.security_group.id as an input parameter into my elb_module. This means terraform must know to create the "security_group_module" first and have its outputs computed so the compute values are passed into the "elb_module", but that does not happen. I get a failure which appears to indicate the "security_group_module" is not computed yet when "elb_module" is trying to be instantiated.

Thanks.

**Comment by apparentlymart** [comment]
@dtserekhman-starz from what you have described here it sounds like you're passing the security group id to the ELB module and thus Terraform should, as you assumed, be able to infer that dependency and create things in the correct order. If that's not working, that sounds like a bug distinct from what this issue is talking about and so I'd encourage you to open a new top-level issue with some config examples so we can dig into it some more. Thanks for sharing this problem!

**Comment by dtserekhman-starz** [comment]
@apparentlymart ,

Just to clarify, here’s how my module is defined which creates a security_group which it then exposes through its output variable below:

module "ecs_elb_security_group" {
  source = "git::ssh://<lib-repository-for-general-resources>//security_group?ref=feature/ecs_v2"

  name="nodify-elb"
  description="Security Group for ECS ELB"

  environment_name="${var.environment_name}"
  environment_type="${var.environment_type}"
  vpc_id="${var.vpc_id}"
}

output "id" {
  value = "${module.ecs_elb_security_group.id}"
}

And this is instantiation of another module which relies on the computed ecs_elb_security_group.id value from above:

module "ecs_elb" {
  source = "git::ssh://<lib-repository-for-general-resources>//classic_load_balancer?ref=feature/ecs_v2"

  environment_name="${var.environment_name}"
  environment_type="${var.environment_type}"

  elb_name="nodify-ecs"
  aws_region="${var.aws_region}"
  subnets="${var.ecs_elb_subnets}"

  //This will be internet facing service
  internal = false

  listeners = [
    {
      lb_port = "${var.nodify_elb_port}"
      lb_protocol = "tcp"
      instance_port = "${var.nodify_docker_host_port}"
      instance_protocol = "tcp"
    }
  ]

  healthcheck_target="HTTP:${var.nodify_elb_port}/"

  security_groups = "${module.ecs_elb_security_group.id}"
}

So, from above, “"${module.ecs_elb_security_group.id}"” value should already be computed when passed to security_groups of the “ecs_elb” module, but it’s not. I believe this does fall into the same realm as other use cases in this ticket? Terraform needs to be able to compute that “ecs_elb_security_group” module is a dependency to “ecs_elb” module and should already instantiate all resources for the “ecs_elb_security_group” before instantiating “ecs_elb” module.

Thanks.

**Comment by apparentlymart** [comment]
@dtserekhman-starz that does sound like a bug, separate from this feature request. If you could open a new issue with these details that'd be very helpful to try to figure out what's going on there.

**Comment by dtserekhman-starz** [comment]
@apparentlymart , I've just opened https://github.com/terraform-providers/terraform-provider-aws/issues/1066 on this.

Thanks!

**Comment by chrisrlong** [comment]
Hi @apparentlymart, We have not actually got to deploy consol yet, <watch this space> currently we just have terraform, chef and capistrano. Terraform "sticks" it together itself using Route 53. Longs we stick to the naming conventions.

**Comment by thomasbiddle** [comment]
This has been a requested feature for about 2 years now according to the other ticket; any progress on this?

**Comment by tomstachura** [comment]
Seems like enough people need this. It's an important feature to be able to manage modules depending on other modules.
In our case we use a variable from module 1 in the provisioner of module 2 and terraform does not detect that. I am thinking that I shouldn't even need depends_on, but I was hoping to patch that problem with it.

**Comment by fpc-jpd** [comment]
+1

**Comment by jpdoria** [comment]
+1

**Comment by jfgibbins** [comment]
If you're looking for additional use cases, that require something like depends_on to work in a module, here's a simple one.  Several infrastructure servers need to be up and running, such as puppet, consul/vault, docker, before other servers that depend on them can even be spun up.  Because of several limitations in terraform resources, modules are the only way to implement configurations in a concise and efficient  manner, versus lengthy and numerous individual resource configurations.  Adding depends_on, cuts my code from 100+ files and thousands of lines of code to 1 file and a couple hundred lines that that have to do nothing more than specify the variables for modules, and a dozen files for the called modules that become a one stop edit for configurations.

**Comment by hmarcelodn** [comment]
Any update on this?

**Comment by dmikalova** [comment]
@apparentlymart Do you expect that the [aws_ecs_service depends_on race condition](https://www.terraform.io/docs/providers/aws/r/ecs_service.html) will be satisfied with cross module variables?

ie `moduleA` outputs `aws_iam_role_policy.moduleA.id` and `moduleB` is declared using that output - is that enough for a `aws_ecs_service` in `moduleB` to not have a race condition, or is a `null_resource` necessary?

**Comment by ozbillwang** [comment]
In my case, I need set `depends_on` to a resource in a module, any work around for me?

```
module "app" {
   depends_on = ["aws_instance.web"]
}
```

I got the same error:

`* module root: module app: depends_on is not a valid parameter`

**Comment by thomasbiddle** [comment]
@ozbillwang You shouldn't need any `depends_on` setup here. Looks like you're referencing your variable incorrectly inside your module.

**Comment by ozbillwang** [comment]
@thomasbiddle 

Thanks for helping, but my codes are fine. Currently I fixed it with below codes

```
data "template_file" "family" {
  template = "${aws_ecs_task_definition.app.family}"

  depends_on = ["aws_ecs_task_definition.app"]
}

data "template_file" "arn" {
  template = "${aws_ecs_task_definition.app.arn}"

  depends_on = ["aws_ecs_task_definition.app"]
}

module "app" {
   source = "./module-xxxx"

   aws_ecs_task_definition_family    = "${data.template_file.family.rendered}"
   aws_ecs_task_definition_arn       = "${data.template_file.arn.rendered}"
}
```

It works, but every time, when run `terraform plan`, always report there is one resource `module.app.aws_ecs_service.app` to be changed. 

But when you apply it, no change at all.

**Comment by FernandoMiguel** [comment]
@apparentlymart in my case, I have a module that creates security groups and then calls it again to create the rules for each SG, passing the corresponding SG_id.
Some SG rules reference other SG_id for inbound targets.

the code that creates the rules depends on the SG already existing.

**Comment by rafaelmagu** [comment]
I'm seeing a very similar issue, but am trying to reference a module output in a resource in the root:
```hcl
# creates one mount target per subnet
resource "aws_efs_mount_target" "web" {
  count           = "${length(module.web-vpc.internal_subnets)}"
  file_system_id  = "${aws_efs_file_system.web.id}"
  subnet_id       = "${element(module.web-vpc.internal_subnets, count.index)}"
  security_groups = ["${aws_security_group.efs.id}"]
}
```
`internal_subnets` is a list.

Attempting to `plan` this will render the following error:
```
Error running plan: 1 error(s) occurred:

* aws_efs_mount_target.web: aws_efs_mount_target.web: value of 'count' cannot be computed
```

**Comment by rafaelmagu** [comment]
I was able to work around that by changing `count` to *3*, `plan/apply`-ing, then reverting the count to the dynamic value.

**Comment by GeoffMillerAZ** [comment]
# Another part to this: 
Consider each of these modules is for user management of accounts. the parent module creates groups and roles and allows users in this account to assume roles in other accounts. It takes a list of the children account IDs to setup allowing this. The child modules create roles with no groups. This is because I want to sign into the parent account and assume roles into other accounts controlled by what group you are placed in the parent account. When you run `terraform plan` you get:

# Error output:
```
Error refreshing state: 1 error(s) occurred:

* module.child_b.provider.aws.${var.iam_alias}-${local.region}: The role "arn:aws:iam::999fakeaccount999:role/myrole" cannot be assumed.
```

There is a provider in the child module that assumes the myrole in the child account. But it can't until the module.parent allows it to. It works fine if I comment out the child module then run it. This is why the error only references child_b and not child_a; I did the commenting out workaround to get it to work for child_a. But due to #9653 I can't run a destroy on these either without manual intervention. So, I'm trying to make sure that permissions checks are taken into consideration with the addition of depends_on to modules. It would also be nice if you could pass a map into a provider key of a module that way I could define my providers outside the module and have them specified by map keys inside the module but that's a side note.

# Snippet of my code
```terraform
// local.config is a non-flat map that holds flags and account numbers 
// and default regions and stuff

module "parent" { // creates groups that allow assume role into other accounts
  source    = "./parent"
  config    = "${local.config}"
  children  = "${values(local.children_map)}"
  iam_alias = "repay_testing_main"
  role      = "myrole"
}

module "child_a" {
  source           = "./child"
  config           = "${local.config}"
  parent_iam_alias = "${module.parent.iam_alias}"
  iam_alias        = "two"
  role             = "myrole"
}

module "child_b" {
  source           = "./child"
  config           = "${local.config}"
  parent_iam_alias = "${module.parent.iam_alias}"
  iam_alias        = "three"
  role             = "myrole"
}
```

**Comment by bluemalkin** [comment]
Any update on implementing this feature please ?

**Comment by melchit20** [comment]
I need this plz

**Comment by gaui** [comment]
Can't resources have `depends_on` on other resources in different modules? I understand that modules can't depend on other modules, but resources should be able to depend on resources, despite of modules.

**Comment by pixelicous** [comment]
+1

Same problems here, again i need to work around terraform to get it to do what i want..
Tried everything, output, direct reference, nothing works.. Even if it to be working with the output, there is no reason to output everything i want to reference, thats just so much overhead. Too much W.E.T code

**Comment by pixelicous** [comment]
#16983 Is this the same error?? This is causing us issues.. do not understand why i cannot just depend on resources from other modules as well, why do we need the output, let me tag resources as public or private. It doesn't work using output anyways..

**Comment by ThomasObenaus** [comment]
## my use-case
I want to use the output of one module (a list) in order to generate a number of resources. The number of resources to be created should match the size of the modules list.

## code snippet
#### module bar
This module provides just a list of sg_ids. One is of a generated resource and one is hardcoded.
```
output "sg_ids" {
  value = [
    "${aws_security_group.sg_bar.id}",
    "sg-b91c79d3",
  ]
}

data "aws_vpc" "default" {
  default = true
}

resource "aws_security_group" "sg_bar" {
  vpc_id      = "${data.aws_vpc.default.id}"
  name        = "sg_bar"
  description = "bar sg"
}
```
#### using module bar
```
data "aws_vpc" "default" {
  default = true
}

provider "aws" {
  region  = "eu-central-1"
  profile = "playground"
}

# module that exports a list of security-groups
module "bar" {
  source = "../bar"
}

resource "aws_security_group" "sg_foo" {
  vpc_id      = "${data.aws_vpc.default.id}"
  name        = "sg_foo"
  description = "foo sg"
}

# add one sg_rule to for each of the security-groups of module bar
resource "aws_security_group_rule" "sg_foo_rule" {
  count                    = "${length(module.bar.sg_ids)}"
  type                     = "ingress"
  from_port                = "8080"
  to_port                  = "8080"
  protocol                 = "tcp"
  source_security_group_id = "${element(module.bar.sg_ids,count.index)}"
  security_group_id        = "${aws_security_group.sg_foo.id}"
}
```
#### Error output
```
------------------------------------------------------------------------

Error: Error running plan: 1 error(s) occurred:

* aws_security_group_rule.sg_foo_rule: aws_security_group_rule.sg_foo_rule: value of 'count' cannot be computed
```

**Comment by sgendler-stem** [comment]
Another use-case

I have a module which creates an ECS service, sets up autoscaling, cloudwatch alarms, etc.  It takes the task definition as an input, which tends to be where the differences between similar services lie.  Different services need different access to other AWS resources, however, which means I need to create a task role and assign policies to it.  In most contexts, I would have the service module define the role and return the arn and name as outputs, to have policies attached after the module returns.  However, because it may sometimes be necessary to have certain policies in place BEFORE the service starts up (to access a KMS key to decrypt secrets in the config file, for example), and the service module waits for the service to start.  For this reason, I need to create the role, attach various policies, and then run the service module.  However, creating the role is a separate act from attaching a policy to it, so the service module ends up dependent on the role because the role arn is passed in, but it starts executing before the policy attachment resource has completed.  It would be easy enough to make the service module depend on the policy attachment resource, except that isn't possible.  So I have to workaround it by passing an output of the attachment resource to the module.  I'm sure the race between policy attachment and actually needing it will almost always work in my favour, but it will inevitably not be guaranteed and it WILL happen, probably the first time I'm not around to babysit a service launch in-person.

```
# Have to define the task role and pass it to the stem-job module because the
# job may not be able to start up if permissions aren't granted first, so we
# cannot attach permissions to the role after the service is started in the
# module
data "aws_iam_policy_document" "ecs_task_role" {
  statement {
    effect = "Allow"
    actions = ["sts:AssumeRole"]
    
    principals {
      type = "Service"
      identifiers = ["ecs-tasks.amazonaws.com"]
    }
  }
}

resource "aws_iam_role" "ecs_task_role" {
  name = "${var.service_name}-${var.environment}-task-role"
  assume_role_policy = "${data.aws_iam_policy_document.ecs_task_role.json}"
}

# attach a policy to grant access to some resource
resource "aws_iam_role_policy_attachment" "test-attach" {
    role       = "${aws_iam_role.ecs_task_role.name}"
    policy_arn = "${aws_iam_policy.policy.arn}"
}

module "job" {
  source = "../../modules/stem-job"

  service_name = "${var.service_name}"
  environment = "${var.environment}"
  ecs_cluster_name = "${data.terraform_remote_state.ecs.ecs_cluster_name}"
  ecs_cluster_arn = "${data.terraform_remote_state.ecs.ecs_cluster_arn}"
  container_definition = "${data.template_file.container_definition.rendered}"

  # IMPORTANT: see comment on depends_on field
  ecs_task_role_arn = "${aws_iam_role.ecs_task_role.arn}"

  # If other policies are added to the task role in separate aws_iam_role_policy
  # resources, add them as dependencies here, so that the module doesn't get
  # instantiated until after the role is populated
  depends_on = ["aws_iam_role_policy_attachment.test-attach"]

  desired_number_of_tasks = "${var.desired_number_of_tasks}"
  min_number_of_tasks = "${var.min_number_of_tasks}"
  max_number_of_tasks = "${var.max_number_of_tasks}"
}
```

**Comment by rvaidya** [comment]
I am using modules simply to reduce the amount of boilerplate I need to write to set up an instance template, a managed instance group, etc. for my machine clusters.  I want to be able to pass in additional resources like load balancer pools, disk images, etc. but I can't set my module to depend on these passed in resources.

**Comment by ozbillwang** [comment]
@sgendler-stem 

In your sample, you can move the codes about iam role, policy to that module as well. 

I did that without `depend_on` issue.

**Comment by pixelicous** [comment]
So....... this is open for a very long time?
I cannot understand why one module cannot depend on another.

I have a module that creates webapps, it needs to receive some items from a different module.. how do i go about doing that??

So basic..

**Comment by agnelantony2** [comment]
is this still an open issue ? any feature is being worked on for this ask ?

**Comment by pb0101** [comment]
This is so basic. I am running into issues with some large stacks. Hopefully this issue gets fixed soon.

**Comment by flmmartins** [comment]
Since it's unlock. Does that means that this issue was implemented?

**Comment by zygimantas** [comment]
My case: two separate modules: VPC and ALB.
In VPC module, VPC and subnets are created. Output variable is "vpc_id".
ALB module has vpc_id variable and use aws_subnet_ids data source to find subnets for ALB.

Unfortunatelly, that does not work. I suspect that data source is evaluated once VPC is created, but before subnets are created. My workaround was to return subnets as output variable from VPC module, and that is not an elegant solution.

**Comment by jcarlson** [comment]
@zygimantas I believe yours is a use case you can solve today by adding `depends_on` attributes to your output blocks.

You VPC module could look something like this:

```terraform
output “vpc_id” {
  depends_on = [“aws_subnet.foo_subnet”, “aws_subnet.bar_subnet”]
  value      = “${aws_vpc.my_vpc.id}”
}
```

This will prevent your VPC module from exposing the `vpc_id` until the subnets are ready, which should allow your ALB module to lookup the subnets with a data block as you would expect.

**Comment by lukasmrtvy** [comment]
I am trying this:

```
variable "dependencies" {
  type = "list"
  default = []
}
resource "null_resource" "dependency_getter" {
  triggers {
    my_dependencies = "${join(",", var.dependencies)}"
  }
}
data "archive_file" "hash" {
  type        = "zip"
  source_dir = "../deployment"
  output_path = "content.zip"
}
resource "null_resource" "deployment" {
  triggers = {
    src_hash = "${data.archive_file.hash.output_sha}"
  }
  depends_on = [
    "null_resource.dependency_getter",
  ]
  provisioner "local-exec" {
    command = ["deployment ..."]
  }
}
resource "null_resource" "dependency_setter" {
  depends_on = [
    "null_resource.deploy"
  ]
}
output "depended_on" {
  value = "${null_resource.dependency_setter.id}-${timestamp()}"
}
```
but, its is always triggering new resource:
```
-/+ module.mymodule.null_resource.dependency_getter (new resource required)
      id:                       "6712669880425280642" => <computed> (forces new resource)
      triggers.%:               "1" => "1"
      triggers.my_dependencies: "1272235997980015531-2019-05-31T11:56:04Z,5997274410980414987-2019-05-31T11:56:04Z,646220093417855083-2019-05-31T11:56:04Z,95820358528843397-2019-05-31T11:56:04Z,3617892051138713807-2019-05-31T11:56:04Z,7776302212176589864-2019-05-31T11:56:04Z" => "1272235997980015531-2019-05-31T12:09:45Z,5997274410980414987-2019-05-31T12:09:45Z,646220093417855083-2019-05-31T12:09:45Z,95820358528843397-2019-05-31T12:09:45Z,3617892051138713807-2019-05-31T12:09:45Z,7776302212176589864-2019-05-31T12:09:45Z"
```
Any idea?

**Comment by min-w-park** [comment]
any workaround here in case of module/submodule structure? When module depends on submodule?

**Comment by queglay** [comment]
I produced an output (like an id on a null resource) on one module,
and then the other module can be fed that id as a variable from your main template.  This variable can be used in the depends_on list, and as a trigger.

**Comment by alexelshamouty** [comment]
Jesus, this issue is not resolved still! I am facing the same issue.

**Comment by FernandoMiguel** [comment]
> Jesus, this issue is not resolved still! I am facing the same issue.

@alexelshamouty and the outputs dont work for you?

**Comment by alexelshamouty** [comment]
@FernandoMiguel No. It doesn't. 

```
module "database" {
  source = "./acceptance/data-stores/mysql"
  aws_region = "${var.aws_region}"
  db_name = "${var.db_name}"
  db_password = "${var.db_password}"
  db_instance = "${var.db_instance}"
  db_username= "${var.db_username}"
}

module "cluster" {
  source = "./acceptance/services/webserver-cluster"
  aws_region = "${var.aws_region}"
}

```

The first module outputs 

```
output "endpoint" {
  value = "${aws_db_instance.wp_db.endpoint}"
}
```

The cluster module I have: 

```
data "terraform_remote_state" "wp_db" {
  backend = "etcd"
  config = {

,,,,,,
  }
  depends_on = ["aws_db_instance.wp_dp"]
}
```

And this is the result:

```

Error: data.terraform_remote_state.wp_db: resource depends on non-existent resource 'aws_db_instance.wp_dp'
```

**Comment by FernandoMiguel** [comment]
@alexelshamouty `aws_db_instance.wp_dp` doesnt exist in that context.
you cant depend on it at root module level

**Comment by FernandoMiguel** [comment]
what you need to do is pass "something" from the outputs of `database` module to the `cluster` module as an input, creating a dependency on creation

**Comment by alexelshamouty** [comment]
@FernandoMiguel Could you please elaborate with a piece of pseudo code? I have been using terraform only for a couple of days now and I have a hard time imagining what the solution would be like.

**Comment by FernandoMiguel** [comment]
@alexelshamouty you already have it in your example.
you have an output.
use that as the input for the other one. consume it inside it. 
done
profit

**Comment by alexelshamouty** [comment]
@FernandoMiguel 
```
module "cluster" {
  source = "./acceptance/services/webserver-cluster"
  aws_region = "${var.aws_region}"
  dp_endoint = module.database.wp_db.endpoint
}

```

This is throwing another error

```

Error: Error parsing /Users/aelshamouty/Documents/Projects/Terraform/terragrunt/live/acceptance/database/.terragrunt-cache/aSxsE4aMVy4XFhs7rOjE5YfyA3Y/glEYwR24V1Nd5cFLqqxNDZnOmMA/main.tf: At 18:16: Unknown token: 18:16 IDENT module.database.wp_db.endpoint
```

I got this from the documentation:
https://www.terraform.io/docs/configuration/modules.html

No profit.

**Comment by DevOpsFu** [comment]
Your output reference is incorrect. The output from the module called `database` is simply called `endpoint`.


Therefore your output reference is `module.database.endpoint`

**Comment by alexelshamouty** [comment]
```
Error: Error parsing /Users/aelshamouty/Documents/Projects/Terraform/terragrunt/live/acceptance/database/.terragrunt-cache/aSxsE4aMVy4XFhs7rOjE5YfyA3Y/glEYwR24V1Nd5cFLqqxNDZnOmMA/main.tf: At 17:16: Unknown token: 17:16 IDENT output.database.endpoint
```

Tried that as well..

**Comment by DevOpsFu** [comment]
It's hard to diagnose without seeing the module structure. It looks like you're perhaps not passing in the output to the module for the cluster. You'd need something like:

```module "cluster" {
  source = "./acceptance/services/webserver-cluster"
  aws_region = "${var.aws_region}"
  endpoint = "${module.database.endpoint}"
}```

(With `endpoint` as a valid variable within your cluster module). And then consume the endpoint value there. Make sense?

**Comment by alexelshamouty** [comment]
This is now working. 

I thought however that this was the reason for the terraform_remote_state. Is to be able to exchange data between module without creating input streams like this one. 

Is this fixed in any of the new versions?

**Comment by FernandoMiguel** [comment]
@alexelshamouty i would take this from this ticket, as it isnt a problem with tf but your code implementation.
try the maililing list or ideally one of the many slacks were we discuss terraform. 
you can find me on the og-aws and hangops

**Comment by FernandoMiguel** [comment]
@alexelshamouty the state only gets updated when the apply finishes.
you are creating two resources IN THE SAME state... if those were two different statefiles, the 2nd would consume the resource outputs of the 1st

**Comment by alexelshamouty** [comment]
Thanks @FernandoMiguel  and @DevOpsFu!

**Comment by chunkydigits** [comment]
I am experiencing this issue as well, but not sure whether I am making a mistake here or whether the fact that I am using ARM templates as a deployment mechanism for the Logic Apps that I am deploying. 

I have a file structure of: 
- main.tf 
-- shared/main.tf
-- logic-apps/main.tf
-- logic-apps/module1/main.tf
-- logic-apps/module2/main.tf

Main is an orchestration level that is lightweight and creates a Resource Group and just called modules below it.  
Shared is supposed to creates a servicebus, auth rule and api_connection to the service bus for use in the 'another' module. This currently outputs a dependency_id as below. 
```
resource "azurerm_template_deployment" "service_bus_api_connection" {
  depends_on          = ["azurerm_template_deployment.salesforce_api_connection"]
  name                = "${var.connections_servicebus_name}"
  resource_group_name = "${var.resource_group_name}"

  template_body = templatefile("${path.module}/arm/servicebus-api.json", {
    connections_servicebus_name  = "${var.connections_servicebus_name}"
    region                       = "${var.resource_group_location}"
    api_id                       = "${var.servicebus_id}"
    servicebus_connection_string = "${azurerm_servicebus_topic_authorization_rule.auth_rule.primary_connection_string}"
  })

  deployment_mode = "Incremental"
}
output "dependency_id" {
  depends_on = ["service_bus_api_connection", "service_bus"]
  value      = "${azurerm_template_deployment.service_bus_api_connection.id}"
}
```

Main.tf then passes that dependency_id to the logic apps module that then creates some logic apps that should point at the servicebus api connection that was created above. Logic-apps.tf then passes this same variable to module1.tf to create the logic apps relevant to module1. 

The module1.tf is using an Arm template in the same way as above, but has a depends_on the dependency_id that has been passed all the way from shared/main.tf. 

At this point i am getting the following error: 

`Error waiting for deployment: Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." Details=[{"code":"NotFound","message":"{\r\n  \"error\": {\r\n    \"code\": \"ApiConnectionNotFound\",\r\n    \"message\": \"The API connection 'servicebus' could not be found.\"\r\n  }\r\n}"}]`

Is this something that is caused by my use of ARM scripts within the Terraform code and are they not completing in the right timespan to allow the later creation of the logic apps that depend on them? 

As a side note, if i run an Apply on this a second time it works because of the pre-existence on the service bus and the connection to it that can now be used, but not sure the client with accept that as a suitable solution. :-D

**Comment by zkghost** [comment]
Just another use case: 
We have one module that creates a template lambda function, it conditionally includes a separate module that provides default cloudwatch alarms for that lambda. 

What I'd like to do is create a default alarm that watches the log output of the lambda and alerts when occurrences of terms like ERROR and WARN reach certain thresholds.

Issue is that the `cloudwatch_log_group` isn't created when I try to create my `log_metric_filter`, but I can't specify that my `log_metric_filter` `depends_on` my `lambda_function`. It crashes, saying the cloudwatch group doesn't exist yet. Trying to specify that the alarm module depends on the `lambda_function` resource also doesn't work. 

It works if I do everything in a single lambda function though, because I can set dependency between my `log_metric_filter` and my `lambda_function`. I will try the output hack in the mean time tho.

**Comment by deeco** [comment]
So I have a dependancies on the module creating the resource first to retrieve the id, is this the only way to pass on wait for deployment ?
The module will create the resource and then the data must interrogate to retrieve the ids.

the module creates fine, when I add in the data and output it does not recognize resources as not created yet.

module "public_outbound_lb" {
  source = "../../../../modules_az_v2/public_outbound_lb"
  location = "${var.location}"
  resource_group_name = "${local.net_resource_group_name}"
  site = "${local.site}"
}
data "azurerm_lb" "publb" {
  name                = "${format("%s_PUB_LB",local.site)}"
  resource_group_name = "${local.net_resource_group_name}"
  depends_on = "module.public_outbound_lb"
}
data "azurerm_lb_backend_address_pool" "publb" {
  name            = "publbpool"
  loadbalancer_id = "${data.azurerm_lb.publb.id}"
  depends_on = "module.public_outbound_lb"
}
output "backend_address_pool_id" {
  value = "${data.azurerm_lb_backend_address_pool.publb.id}"
  depends_on = "module.public_outbound_lb"
}

**Comment by GavBurke** [comment]
> This is very strange - we introduced depends_on for modules in 0.8.0-beta2
> 
> Paul

I'm not sure that is correct, using 0.11 and a module block cannot contain a depends_on

```
error: module "example": "depends_on" is not a valid argument
```

**Comment by yaman** [comment]
It has been almost 3 years since this issue was opened, and I can verify that there is no way of structuring a complex infrastructure into simple terraform module building blocks that depends on each other(try creating k8s namespaces in one module that needs to be done right after another module that is responsible for GKE node_pool creation). 

_I hope I am missing something here._ 

What do you need for this feature to be implemented? Can we help you? I think some of us could write terraform from scratch in 3 years... Have some faith on the community. @mitchellh

**Comment by mitchellh** [comment]
Hi @yaman! A lot of the work in 0.12.0 was laying the groundwork for this to be possible. We cut 0.12.0 without it so we can ship sooner, but a lot of the features we intended on shipping are being worked on in 0.12.x. For example, we shipped resource `for_each` a couple releases ago.

Kristin (one of our core team here) gave a great talk at HashiConf EU titled "A (2nd) Tour of Terraform 0.12" where she outlines some of the roadmap. She notes that "module depends_on" is on the way (slide 52) and it is our intention to build that into the 0.12.x release series.

https://www.hashicorp.com/resources/a-2nd-tour-of-terraform-0-12

**Comment by yaman** [comment]
> Hi @yaman! A lot of the work in 0.12.0 was laying the groundwork for this to be possible. We cut 0.12.0 without it so we can ship sooner, but a lot of the features we intended on shipping are being worked on in 0.12.x. For example, we shipped resource `for_each` a couple releases ago.
> 
> Kristin (one of our core team here) gave a great talk at HashiConf EU titled "A (2nd) Tour of Terraform 0.12" where she outlines some of the roadmap. She notes that "module depends_on" is on the way (slide 52) and it is our intention to build that into the 0.12.x release series.
> 
> https://www.hashicorp.com/resources/a-2nd-tour-of-terraform-0-12

Thanks for the prompt reply @mitchellh, following release notes closely ;)

**Comment by JnMik** [comment]
Managing to spawn the different modules using a dependency workaround is possible, yet it gets really messy when you want to destroy the stack. (Cycle errors happens and they are hard to read, even if there is no real cycle error in the module dependencies for what I know). 

IMO this feature is a must to do what I imagine of "clean architecture code" with Terraform for large stack. I still need some shell script to deploy / destroy the modules in the correct order because with every workaround I could find on the net I never managed to have a fully working solution.

Keep up the good work TF Team we're waiting for this ! :1st_place_medal:

**Comment by ConscriptMR** [comment]
is this planned to be solved?

**Comment by scarolan** [comment]
Anyone have a decent workaround for doing this in Terraform 0.11.x?  My use case is simple, I want the module to depend on a resource that must be created first before the module does anything.

**Comment by JamesDLD** [comment]
Don't know if a workaround could be decent :D but you you can output a resource and fill this output as input of your module.

**Comment by khanali21** [comment]
Any time frame for this? Are we expecting this to be fixed soon?

**Comment by jspiro** [comment]
Likewise, not having depends_on with modules, nor for_each, discourages use of modules in terraform since you have to give up critical features to take advantage.

**Comment by iniinikoski** [comment]
I heard an update weeks ago in a Hashicorp event telling us "soon". :) Maybe Christmas comes early this year ;)

**Comment by buamod** [comment]
I really hope so .

**Comment by jarbread** [comment]
hope this gets implemented soon.

**Comment by ortelli-oxalide** [comment]
Hi guys, any release ETA for this so much expected feature ?

**Comment by danieldreier** [comment]
@ortelli-oxalide we just picked this up and started working on it. Two engineers on the terraform core team recently started working on supporting count, for_each, and depends_on work in modules. 

Modules should be first-class, and the message is really clear from the terraform community that having some language features not work in modules makes it hard to adopt modules as intended, which has downstream consequences like making it harder to write maintainable, well organized terraform code. We saw how many people asked for it, we understand why not having it is painful, have prioritized it, and are actively working on it.

This is one of the terraform core team's current top priorities, and it's also a pretty significant amount of work, more than you'd imagine, because it's part of a larger effort to make modules first-class. The project just started, and I don't have an ETA.

As the project is currently running, it looks like for_each and count will be done first, and then depends_on will follow afterward. However, it may be that we batch them all up in the same release to minimize churn.

**Comment by jspiro** [comment]
Thank you @danieldreier – epic update and the transparency is appreciated! I think we can finally all rest easy that this is coming ^_^

**Comment by chrisrlong** [comment]
Thanks is good news @danieldreier. Time will come to remove all the work arounds soon then...

**Comment by jared2501** [comment]
Glad to hear! super exciting

**Comment by ortelli-oxalide** [comment]
Thanks @danieldreier we are looking forward to that ! 

As I cannot see anyone else reporting a specific issue related to module dependencies that I'm encountering, I'll report it here. I hope that the support of "depends_on" on modules will resolve that, but I'm not quite sure. Any idea ?

With some code like this:
```
module "some_module" {
  ...
  some_parameter = module.another_module.output
  ...
}

# some_module source code
resource "some_resource" "_" {
  count = length(var.send_rds_logs_to_dd_with_lambda_arn)
  ...
}
```

Terraform raise the following error (at plan or apply):
```
Error: Invalid count argument
...
in resource "some_resource" "_":
   2:   count = length(var.send_rds_logs_to_dd_with_lambda_arn)

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.
```

**Comment by matti** [comment]
I think I managed to create a valid workaround for this: https://github.com/matti/terraform-module-depends_on

It requires modules to have some boilerplate, but otherwise it seems to work fine.

**Comment by kuwas** [comment]
@danieldreier @apparentlymart Thanks for the update and putting the emphasis on this, it is greatly appreciated.

**Comment by solidDoWant** [comment]
@danieldreier Great to hear you guys are working on this. Have you guys been able to scope and figure out a timeline for the project yet? Roughly when this is planned to be released is one of the deciding factors my team will be using TFC or switching to Pulumi. Can you provide any broad information on when this will be released (week/month/year)?

**Comment by danieldreier** [comment]
@solidDoWant While I appreciate the need for you and others to plan, we are not yet in a position to communicate any timelines. My hope above was to begin communicating what we’re currently focused on more publicly. 

I can give a few details on the larger first-class module project that this is a part of to help you with your planning: 

- **improving modules is the main feature of the next terraform release**: we're expecting to ship  make modules first-class in 0.13, which is the next major release. The only other thing that we'll block 0.13.0 on is a set of critical bug fixes that require backward-incompatible behavior changes. We're not planning on holding back 0.13.0 for any other features.
- **depends_on may ship after count and for_each**: depends_on is a subset of the larger project of first-class modules. As part of that project, we'll make depends_on, count, and for_each work in modules. I can't tell you, yet, whether we'll ship depends_on, and so count and for_each all at once, or what order they'll ship in - we may ship for_each and count in 0.13.0 and then quickly follow up with a minor release that adds depends_on. We do intend to batch any breaking changes in 0.13.0, so that the later addition of depends_on would be purely additive.

**Comment by jspiro** [comment]
That prioritization makes a lot of sense to me. Looking forward to it!

**Comment by pixelicous** [comment]
Great, 3 years in, looking forward to having this needed feature.

**Comment by hakro** [comment]
This is a really important feature to have in my opinion. It would make the code much more readable and maintainable than by using one of the hack that we find on blog posts.

**Comment by emmaLP** [comment]
Can anyone please give a time line on when this feature will be released?

**Comment by max-l-weaver** [comment]
Wow, over three years and still nothing?!

**Comment by matti** [comment]
three years is nothing in terraform time

**Comment by ketzacoatl** [comment]
> Wow, over three years and still nothing?!

@max-l-weaver with this comment, you are showing how unappreciative and ignorant you can be.

You can fault the devs for not getting this right with modules from the start, but aside from that, you should have more awareness of reality and what these changes have required before sharing those types of crass comments.

In fact, you should probably not use any open source software, as the type of comment you left above are one of the major reasons why application developers give up on open source.

Please be more kind and appreciative in your comments, or leave them to yourself.

**Comment by solidDoWant** [comment]
Just because the software is open source doesn't mean we aren't paying for it. Many of us (myself previously included) pay a monthly or yearly fee for Terraform/TFC and in term expect support and timely software updates. The fact that Hashicorp rarely if ever shares roadmaps or any kind of even a guess at a timeline quite frankly makes it unusable for business critical services, such as the infrastructure backing software company's services. The lack of communication on one of, if not the most requested feature of the last 3-4 years is not something to appreciate.

**Comment by confiq** [comment]
@solidDoWant , chillrex man...  Many people "pay a monthly" and wait patiently for a new feature. 
This specific feature requires lot of tech resources and time, no need for such an outcry because it does not help anyone. 

They decided to implement it which is great news, crass comments like these do not help anyone.

**Comment by kuwas** [comment]
@solidDoWant There are various workarounds for this feature that currently exist, such using templates to generate tf files, so its not really a show-stopper. We should be thankful that they even prioritized this at all.

**Comment by solidDoWant** [comment]
To clarify, my issue isn't at all with the lack of the feature - it's with the lack of communication. I would be fine with Hashicorp never developing this feature, but I take issue with Hashicorp not being able to even take a guess as to when it will be completed, if at all. It's great that they've communicated that this is a top priority, as that address the 'if it will happen', but they haven't addressed the 'when'. 

This Github issue isn't the only place affected by the underlying company issue either. There are 136 PRs on this repo alone (not to mention the numerous providers) that are open. Out of those 135, only one has been assigned a milestone. Roughly 15 haven't even been commented on. On most of these there is no indication when, or if these PRs will be addressed. Not great from a user's perspective when there's a fix you're waiting on in one of them, or in a developer's perspective when your PR sits open for months or years.

Hashicorp makes some cool tools and (from what I've seen) most of their code is pretty solid. I've no issues with the software developers, but their product strategy and overall communication leaves a lot to be desired.

**Comment by camjackson** [comment]
@danieldreier can you please lock this issue to contributors only? I believe at this point just about every *constructive* comment that could be made, has already been made.

There are likely hundreds of people subscribed to this issue, and we don't all need to get an email every time someone wants to complain. I'll be happy if the next time I hear about this feature is when it's done.

**Comment by ketzacoatl** [comment]
> To clarify, my issue isn't at all with the lack of the feature - it's with the lack of communication.

Thanks for clarifying.

> I would be fine with Hashicorp never developing this feature, but I take issue with Hashicorp not being able to even take a guess as to when it will be completed, if at all. It's great that they've communicated that this is a top priority, as that address the 'if it will happen', but they haven't addressed the 'when'.

I can understand that frustration, I agree Hashicorp could improve their communication with us as users of various kinds.

However, WRT this general topic of first-class modules, I believe Hashicorp has been pretty clear, and within the bounds of what's possible (and also avoiding making unnecessary expectations which they may fail to deliver on).

For example, in the 3 years that this has transpired, we've seen the following:
1) many, rather verbose and elaborate discussions of the use cases and needs and possibilities, the problems and blocks, as well as how to get there.
2) Hashicorp realizes their fsck up, and acknowledges that TF core needs to be refactored in major ways.
3) Hashicorp lays down the roadmaps for 0.11 and 0.12, with various breaking/major changes that will facilitate the next round of changes
4) Hashicorp plows through 0.11 and 0.12 releases, following through as planned.
5) Hashicorp starts work on 0.13 which is focusing on modules as a first-class citizen

Particularly for those last two,  no timelines (with unreasonable expectations) were laid down. Yea, I would have loved to see 0.12 completed sooner, or dev on 0.13 to have started sooner, but I'd rather this than to have a product which is all over the place, poorly developed, poorly designed, lots of major breaking changes, and just generally moving to the whims of the wind as it blows.

**Comment by max-l-weaver** [comment]
@ketzacoatl please take a deep breath and calm down.

Imo Hashicorp market themselves as a software provider that provides solutions for businesses. If it was some dude working on his own in his basement then of course i wouldn't post such ignorant comments, however the fact that this feature has been open for over three years with no communication is a little disconcerting. I apologise if you felt somehow attacked by my comment.

**Comment by phildier** [comment]
Please take the complaints to twitter or reddit.  I'm subscribed to this thread for news about the feature only.

**Comment by danieldreier** [comment]
Based on the feedback here, I've locked this discussion. We're actively working on module expansion. The engineers are making good progress, and I am excited about 0.13. For folks who are really upset, I'm ddreier@hashicorp.com and I'm happy to talk.

**Comment by danieldreier** [comment]
I'm very excited to announce that beta 1 of terraform 0.13.0 will be available on June 3rd. I'm expecting this feature to be in that release. I've [pinned an issue](https://github.com/hashicorp/terraform/issues/25016) with more details about the beta program, and posted a [discuss thread](https://discuss.hashicorp.com/t/terraform-v0-13-0-beta-program/9066) for folks who want to talk about it more.
